protocol/java/org/openyolo/protocol/CredentialSaveRequest.java

```diff
@@ -0,0 +1,244 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.openyolo.protocol;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+import android.support.annotation.NonNull;
+
+import org.openyolo.protocol.internal.ByteStringConverters;
+import org.openyolo.protocol.internal.ClientVersionUtil;
+import org.openyolo.protocol.internal.CollectionConverter;
+import org.openyolo.protocol.internal.CredentialConverter;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.hamcrest.core.Every.everyItem;
+import static org.hamcrest.core.IsNull.notNullValue;
+import static org.openyolo.protocol.internal.CustomMatchers.notNullOrEmptyString;
+import static org.valid4j.Assertive.require;
+
+/**
+ * A request to save a given {@link Credential credential}. To make the request the client
+ * encodes it in its associated protocol buffer byte array form in an activity intent
+ * via the {@link ProtocolConstants#EXTRA_SAVE_REQUEST} extra.
+ *
+ * @see <a href="https://spec.openyolo.org/openyolo-android-spec.html#saving-credentials">
+ *     OpenYOLO specification: Saving Credentials</a>
+ */
+public final class CredentialSaveRequest implements Parcelable {
+
+    /**
+     * Parcelable reader for {@link CredentialSaveRequest} instances.
+     * @see android.os.Parcelable
+     */
+    public static final Creator<CredentialSaveRequest> CREATOR = new CredentialSaveRequestCreator();
+
+    @NonNull
+    private final Credential mCredential;
+
+    @NonNull
+    private final Map<String, byte[]> mAdditionalProperties;
+
+    /**
+     * Reads a save request from its protocol buffer byte array form.
+     * @throws IOException if the given save request could not be decoded.
+     *
+     * @see #toProtocolBuffer()
+     */
+    public static CredentialSaveRequest fromProtoBytes(byte[] credentialSaveRequestBytes)
+            throws IOException {
+        if (credentialSaveRequestBytes == null) {
+            throw new IOException("Unable to decode credential save request from null array");
+        }
+
+        return new CredentialSaveRequest.Builder(
+                Protobufs.CredentialSaveRequest.parseFrom(credentialSaveRequestBytes))
+                .build();
+    }
+
+    /**
+     * Creates a {@link CredentialSaveRequest} from the given {@link Credential}.
+     */
+    public static CredentialSaveRequest fromCredential(Credential credential) {
+        return new Builder(credential).build();
+    }
+
+    private CredentialSaveRequest(Builder builder) {
+        mCredential = builder.mCredential;
+        mAdditionalProperties = builder.mAdditionalProps;

```

Comment:
To ensure immutability:
`mAdditionalProperties = Collections.unmodifiableMap(builder.mAdditionalProps)`